unison-sync/src/Unison/Sync/Types.hs

```diff
@@ -0,0 +1,506 @@
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON DownloadEntitiesRequest where
+  toJSON (DownloadEntitiesRequest repoName hashes) =
+    object
+      [ "repo_name" .= repoName,
+        "hashes" .= hashes
+      ]
+
+instance FromJSON DownloadEntitiesRequest where
+  parseJSON = Aeson.withObject "DownloadEntitiesRequest" $ \obj -> do
+    repoName <- obj .: "repo_name"
+    hashes <- obj .: "hashes"
+    pure DownloadEntitiesRequest {..}
+
+data DownloadEntitiesResponse = DownloadEntitiesResponse
+  { entities :: NEMap Hash (Entity HashJWT Hash Text)
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON DownloadEntitiesResponse where
+  toJSON (DownloadEntitiesResponse entities) =
+    object
+      [ "entities" .= entities
+      ]
+
+data UpdatePathRequest = UpdatePathRequest
+  { path :: RepoPath,
+    expectedHash :: Maybe TypedHash, -- Nothing requires empty history at destination
+    newHash :: TypedHash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON UpdatePathRequest where
+  toJSON (UpdatePathRequest path expectedHash newHash) =
+    object
+      [ "path" .= path,
+        "expected_hash" .= expectedHash,
+        "new_hash" .= newHash
+      ]
+
+instance FromJSON UpdatePathRequest where
+  parseJSON = Aeson.withObject "UpdatePathRequest" $ \obj -> do
+    path <- obj .: "path"
+    expectedHash <- obj .: "expected_hash"
+    newHash <- obj .: "new_hash"
+    pure UpdatePathRequest {..}
+
+-- | Not used in the servant API, but is a useful return type for clients to use.
+data UpdatePathResponse
+  = OutOfDate HashMismatch
+  | MissingDependencies (NeedDependencies Hash)
+  deriving stock (Show, Eq, Ord)
+
+data NeedDependencies hash = NeedDependencies
+  { missingDependencies :: NESet hash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON hash => ToJSON (NeedDependencies hash) where
+  toJSON (NeedDependencies missingDependencies) =
+    object ["missing_dependencies" .= missingDependencies]
+
+instance (FromJSON hash, Ord hash) => FromJSON (NeedDependencies hash) where
+  parseJSON = Aeson.withObject "NeedDependencies" $ \obj -> do
+    missingDependencies <- obj .: "missing_dependencies"
+    pure NeedDependencies {..}
+
+data HashMismatch = HashMismatch
+  { repoPath :: RepoPath,
+    expectedHash :: Maybe TypedHash,
+    actualHash :: Maybe TypedHash
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON HashMismatch where
+  toJSON (HashMismatch repoPath expectedHash actualHash) =
+    object
+      [ "repo_path" .= repoPath,
+        "expected_hash" .= expectedHash,
+        "actual_hash" .= actualHash
+      ]
+
+instance FromJSON HashMismatch where
+  parseJSON = Aeson.withObject "HashMismatch" $ \obj -> do
+    repoPath <- obj .: "repo_path"
+    expectedHash <- obj .: "expected_hash"
+    actualHash <- obj .: "actual_hash"
+    pure HashMismatch {..}
+
+data UploadEntitiesRequest = UploadEntitiesRequest
+  { repoName :: RepoName,
+    entities :: Map Hash (Entity TypedHash TypedHash Text)
```

Comment:
Admittedly I don't know whether it's worth it, but if we are embracing NonEmpty, then we'll want it here too.